src/cider/type/plan/FunctionExpr.cpp

```diff
@@ -0,0 +1,295 @@
+}
+
+JITExprValue& FunctionOper::codegen(CodegenContext& context) {
+  JITFunction& func = *context.getJITFunction();
+  ExtensionFunction ext_func_sig = [=]() {
+    try {
+      return bind_function(this);
+    } catch (CiderCompileException& e) {
+      LOG(WARNING) << "codegenFunctionOper[CPU]: " << e.what();
+      throw;
+    }
+  }();
+  const auto& ret_ti = this->get_type_info();
+  CHECK(ret_ti.is_integer() || ret_ti.is_fp() || ret_ti.is_boolean());
+  auto ret_ty = ext_arg_type_to_JIT_type_tag(ext_func_sig.getRet());
+  std::vector<JITValuePointer> arg_lv_values;
+  std::vector<JITValuePointer> arg_lv_nulls;
+
+  // first call
+  if (!is_rewritten_) {
+    const auto& ext_func_args = ext_func_sig.getArgs();
+    for (size_t i = 0; i < this->getArity(); ++i) {
+      // target arg type
+      const auto ext_func_arg = ext_func_args[i];
+      const auto arg_target_ti = ext_arg_type_to_type_info(ext_func_arg);
+
+      // origin arg type
+      const auto arg = this->getArg(i);
+      const auto arg_ptr = this->getOwnArg(i);
+      const auto& arg_ti = arg->get_type_info();
+
+      // Arguments must be converted to the types the extension function can handle.
+      Analyzer::Expr* arg_expr = nullptr;
+      // need cast
+      if (arg_ti.get_type() != arg_target_ti.get_type()) {
+        auto arg_casted_ptr = expr_rewrite_to_cast(arg_ptr, arg_target_ti.get_type());
+        arg_expr = arg_casted_ptr.get();
+        rewrote_args_.push_back(arg_casted_ptr);
+      } else {
+        arg_expr = const_cast<Analyzer::Expr*>(arg);
+        rewrote_args_.push_back(arg_ptr);
+      }
+      FixSizeJITExprValue arg_lv_fixedsize(arg_expr->codegen(context));
+      arg_lv_values.emplace_back(arg_lv_fixedsize.getValue());
+      arg_lv_nulls.emplace_back(arg_lv_fixedsize.getNull());
+    }
+  } else {
+    // second call
+    CHECK_EQ(this->getArity(), rewrote_args_.size());
+    for (size_t i = 0; i < this->getArity(); ++i) {
+      FixSizeJITExprValue arg_lv_fixedsize(rewrote_args_[i]->codegen(context));
+      arg_lv_values.emplace_back(arg_lv_fixedsize.getValue());
+      arg_lv_nulls.emplace_back(arg_lv_fixedsize.getNull());
+    }
+  }
+
+  // get null value
+  auto null = func.createVariable(JITTypeTag::BOOL, "ret_null", false);
+  bool is_nullable = ext_func_call_requires_nullcheck(this);
+  // null is true when at least one argument is null.
+  if (is_nullable) {
+    null = codegenFunctionOperNullArgForArrow(func, this, arg_lv_nulls);
+  }
+
+  // Cast the return of the extension function to match the FunctionOper
+  if (!(ret_ti.is_buffer()) && !is_rewritten_) {
+    const auto extension_ret_ti = ext_arg_type_to_type_info(ext_func_sig.getRet());
+    if (is_nullable && extension_ret_ti.get_type() != this->get_type_info().get_type()) {
+      is_rewritten_ = true;
+      std::shared_ptr<Analyzer::Expr> this_ptr(nullptr);
+      this_ptr.reset(this);
+      auto ret = expr_rewrite_to_cast(this_ptr, this->get_type_info().get_type())
```

Comment:
why not pass `this` directly?

Reply:
the constructor of `Analyzer::UOper` needs a `shared_ptr` as input.